### DISCUSSION TITLE
Silent Refresh IFrame explicitly forbidden?

### DISCUSSION THREAD

**Discussion by dvv** [kanidm/kanidm#2519] [Q&A] [answered]
Hi!

I'm trying to employ "Silent Refresh IFrame" technique and the process chokes with:
```
Refused to frame 'https://idm.example.com/' because an ancestor violates the following Content Security Policy directive: "frame-ancestors 'none'".
```
in the browser.
I could completely override the header but it contains some sha-s so it wouldn't be easy and robust.
I tried to tweak with nginx:
```
add_header Content-Security-Policy "; frame-ancestors https://foo.example.com";
```
but the browser didn't buy that.

I wonder what could be a solution?

TIA

**Comment by yaleman** [comment, marked as answer]
Embedding IDPs in iFrames is a great way to get compromised, that's why we block it.

**Comment by Firstyear** [comment]
The better question here is "what are you trying to achieve" by having these frames.

**Reply by dvv** [reply]
We're exploring the use cases and were trying to follow https://github.com/manfredsteyer/angular-oauth2-oidc/blob/master/docs-src/silent-refresh.md

**Reply by Firstyear** [reply]
So the first thing I notice here is this is for the implicit flow, which we don't support - we only do authorisation code flow. This is actually discussed here https://github.com/manfredsteyer/angular-oauth2-oidc/blob/master/docs-src/token-refresh.md with the "this.oauthService.refreshToken()" for SPAs. Can you attempt with this api instead?

**Reply by dvv** [reply]
I did. We have switched to authorisation code flow.

**Reply by Firstyear** [reply]
Okay, does that api work without the iframe being embedded?

**Reply by dvv** [reply]
I does. The only quirk (?) I encountered is that vanilla `this.oauthService.refreshToken()` will choke [here](https://github.com/kanidm/kanidm/blob/a315d8d440712dafcd7844caacd62c4c7492a6f5/server/lib/src/idm/oauth2.rs#L1164).
The client should pass all the scopes he received, not the scopes he requested.
This is due to supplemental scopes exist in kanidm and one may request "openid email" scope and receive "openid email can_write foo bar" scope.
I spent hours until underestood that, because I didn't know the cause.
I believe we should explicitly state [here](https://github.com/kanidm/kanidm/blob/a315d8d440712dafcd7844caacd62c4c7492a6f5/server/lib/src/idm/oauth2.rs#L1163) what is requested and what is expected.
TIA

**Reply by Firstyear** [reply]
That is a good point, https://www.rfc-editor.org/rfc/rfc6749#section-6 does say it should only contain scopes that were issued, not what was requested. 

I think this is an issue we should open and resolve.

**Reply by dvv** [reply]
Will you take care of properly filing the issue?

**Reply by Firstyear** [reply]
https://github.com/kanidm/kanidm/issues/2527

**Reply by Firstyear** [reply]
Sorry, really busy today, so I won't get to solving it for a few days.

**Reply by dvv** [reply]
Thank you! It's not pressing (as personally _I do know_ the way out) at all.

**Reply by Firstyear** [reply]
Yeah, in the mean time you can send the refresh with no scopes requested and it will work.
